### DISCUSSION TITLE
ESP32 controlled by AT commands, limitations

### DISCUSSION THREAD

**Discussion by dacattack** [espressif/arduino-esp32#9792] [General]
I am currently working on a project where an ESP32-WROVER module is controlled by another microcontroller, specifically an STM32g070 (less capabilities and no RTOS), via AT commands over an RS232 interface. My question are related to the limitations of this setup, particularly in the context of utilizing the advanced features of the ESP32. The way I see it, it makes no sense to use a way less powerful MCU like the STM32 as the master and leave the more powerful and with more capabilities ESP32 as an auxiliary module, controlled by AT commands via RS232. Here are my doubts:

Real-Time Operating System (RTOS): Is it possible to leverage FreeRTOS on the ESP32 while it is operating under AT command control? Can the RTOS functionalities be effectively utilized in this configuration? 

Multicore Processing: The ESP32 is a dual-core microcontroller. Can both cores be efficiently used when the ESP32 is controlled via AT commands? Are there specific commands or configurations that enable this, or are we restricted to single-core operation? Is it even possible?

Real-Time Tasks: How are real-time tasks managed when the ESP32 is controlled through AT commands? Is there a way to prioritize tasks or ensure timely execution similar to what is achievable through direct FreeRTOS programming? Is it even possible?

Thank you in advance for your help!

**Comment by SuGlider** [comment]
> The way I see it, it makes no sense to use a way less powerful MCU like the STM32 as the master and leave the more powerful and with more capabilities ESP32 as an auxiliary module, controlled by AT commands via RS232. 

I agree...

>Real-Time Operating System (RTOS): Is it possible to leverage FreeRTOS on the ESP32 while it is operating under AT command control? Can the RTOS functionalities be effectively utilized in this configuration?

Yes. Note that ESP32 IDF and ESP32 Arduino use FreeRTOS as its base. ESP32 Arduino uses ESP32 IDF as base which has FreeRTOS under it.
The ESP32 AT Command firmware is an ESP32 IDF application that also runs under FreeRTOS.

>Multicore Processing: The ESP32 is a dual-core microcontroller. Can both cores be efficiently used when the ESP32 is controlled via AT commands? Are there specific commands or configurations that enable this, or are we restricted to single-core operation? Is it even possible?

AT Command is an application as written above. FreeRTOS is the one that takes cares of SMP in order to schedule Tasks for both cores, whenever necessary and as designed by the application.

The AT Command Application is open source:
https://github.com/espressif/esp-at/tree/master

>Real-Time Tasks: How are real-time tasks managed when the ESP32 is controlled through AT commands? Is there a way to prioritize tasks or ensure timely execution similar to what is achievable through direct FreeRTOS programming? Is it even possible?

Not sure about how is the design of the AT Command Firwmwarem but you can investigate it based on  its Open Source Code:
https://github.com/espressif/esp-at/tree/master/components/at

As said, FreeRTOS is usually active within ESP32 IDF applications. It is possuble to use FreeRTOS API to create new Tasks, pin the Tasks to specific Core and so on...

**Reply by dacattack** [reply]
Thanks for your explanation, it was really helpful.

**Comment by JAndrassy** [comment]
@dacattack maybe discuss it there https://github.com/espressif/esp-at
